templates/boltzmann/middleware/honeycomb.ts

```diff
@@ -103,24 +158,137 @@ function trace ({
-    return (next: Handler) => (context: Context) => next(context)
-  }
-
+function beelineMiddlewareSpans ({name}: {name?: string} = {}) {
   return function honeycombSpan (next: Handler) {
     return async (context: Context) => {
       const span = beeline.startSpan({
-        name: `mw: ${name}`
+        name: middlewareSpanName(name)
+      })
+
+      beeline.addContext({
+        [otelSemanticConventions.SemanticResourceAttributes.SERVICE_NAME]: honeycomb.options.serviceName,
+        'honeycomb.trace_type': 'beeline'
       })
 
       // Assumption: the invariant middleware between each layer
       // will ensure that no errors are thrown from next().
       const result = await next(context)
+
       beeline.finishSpan(span)
       return result
     }
   }
 }
 
+/*
+ * ┏┓
+ * ┃┃╱╲ in
+ * ┃╱╱╲╲ this
+ * ╱╱╭╮╲╲house
+ * ▔▏┗┛▕▔ we
+ * ╱▔▔▔▔▔▔▔▔▔▔╲
+ * trace with opentelemetry
+ * ╱╱┏┳┓╭╮┏┳┓ ╲╲
+ * ▔▏┗┻┛┃┃┗┻┛▕▔
+ */
+function otelTrace () {
+  return function honeycombTrace (next: Handler) {
+    return (context: Context) => {
+      let traceContext = otel.context.active()
+
+      // Typically, HTTP auto-instrumentation will create a parent span for us
+      let span: otel.Span | undefined = otel.trace.getSpan(traceContext)
+
+      // startSpan should always return a span, but as far as typescript is
+      // concerned the span could still be undefined. We could assert that it
+      // exists, but throwing instrumentation-related errors is poor form.
+      if (span) {
+        span.setAttribute('boltzmann.http.query', context.url.search)
+
+        span.updateName(traceName(context.method, context.url.pathname))
+
+        traceContext = otel.trace.setSpan(traceContext, span)
+
+        if (isDev()) {
+          context._honeycombTrace = span
+        }
+      } else {
+        otel.diag.error(String(new assert.AssertionError({
+          message: 'no parent span found or created',
+          actual: span,
+          expected: true,
+          operator: '=='
+        }).stack))
+      }
+
+      onHeaders(context._response, function () {
+        const handler: Handler = <Handler>context.handler
+
+        if (!span) {
+          return
+        }
+
+        span.setAttribute(
+          otelSemanticConventions.SemanticResourceAttributes.SERVICE_VERSION,
+          <string>handler.version
+        )
+
+        Object.entries(context.params).forEach(([key, value]) => {
+          if (span) {
+            span.setAttribute(paramSpanAttribute(key), value)
+          }
+        })
+      })
+
+      return otel.context.with(traceContext, () => {
+        return next(context)
+      })
+    }
+  }
+}
+
+function otelMiddlewareSpans ({name}: {name?: string} = {}) {
+  return function honeycombSpan (next: Handler) {
+    return async (context: Context) => {
+      let traceContext = otel.context.active()
+
+      const span = honeycomb.tracer.startSpan(
+        middlewareSpanName(name),
+        { kind: otel.SpanKind.SERVER },
```

Comment:
I think that a middleware span should be `SpanKind.INTERNAL` ?

Reply:
Yeah? I saw you linked some of honeycomb's opinions on these - that's helpful, since otel's opinions didn't seem particularly well-voiced. I'll take a look!

Reply:
(when that link talks about asynchronous, they mean like in the queue sense vs RPC sense not async/await sense. Dang terminology collision ugh)

Reply:
I did the investigation here and we already touched base on this understanding, but for documentation purposes: the HTTP top-level trace should be SERVER and outgoing requests should be CLIENT, but everything else should be INTERNAL. The HTTP auto-instrumentation should take care of those first two, and I made the adjustments in the middleware/handler instrumentation. We should be good!